### DISCUSSION TITLE
automatically clean up preserved jobs?

### DISCUSSION THREAD

**Discussion by jrochkind** [bensheldon/good_job#664] [Q&A]
The changelog for 3.0 release says "By default, preserve job records and automatically them clean up"

But I can't find any other documentation of the "automatically clean them up" part.  Perhaps I've just missed this in the long readme though, feel free to direct me to relevant docs! 

Otherwise, can you confirm, are preserved jobs cleaned up automatically now? I looked at the [linked PR](https://github.com/bensheldon/good_job/pull/545), but wasn't sure what was going on in there. Is there configuration to control that, to turn it on or off or control how often the cleanup task runs?  If I have multiple workers running good job, are they all running cleanup tasks? But this isn't anticipated to be a problem?  

Thanks for any info!

**Comment by bensheldon** [comment]
This was the original feature PR, that was released in v2.8, that was enabled by default in v3: https://github.com/bensheldon/good_job/pull/465

(I just linked #545 to that PR too).

Yes, preserved jobs are cleaned up automatically now. Though it looks like the documentation is maybe wrong on the preservation duration because the default configuration says [14 days](https://github.com/bensheldon/good_job/blob/main/lib/good_job/configuration.rb#L20).

**Reply by addisonmartin** [reply]
So, is the correct way in v3 to preserve jobs indefinitely to set:
config.good_job.preserve_job_records = true
config.good_job.cleanup_discarded_jobs = false

**Reply by bensheldon** [reply]
@addisonmartin GoodJob is not architected/intended to preserve job records indefinitely. 

You can achieve something like that with `config.good_job.cleanup_preserved_jobs_before_seconds_ago = Float::INFINITY`

...but I don't recommend treating job records like business records. GoodJob's database tables, while I encourage you to become comfortable with them...are still GoodJob's internal implementation details and subject to change in future versions.

**Reply by jrochkind** [reply]
I would predict performance will also suffer if the job table gets *too* big.

**Reply by addisonmartin** [reply]
Thank you. I understand the negative consequences of preserving them too long, but during development and certain low volume production settings I was interesting in keeping them and seeing when it started to become an issue. Thanks again

**Comment by jrochkind** [comment]
OK, I think I found some of the answer at: 

https://github.com/bensheldon/good_job#monitor-and-preserve-worked-jobs

But I'm getting some conflicting information from code and docs....

* Is default `cleanup_preserved_jobs_before_seconds_ago` 14 days or 1 day?  README says both at different places. Trying to go to the code... it might be 14 days by default when the automatic scheduler runs but 1 day by default when the command line runs? But... both are overridden to the same thing when setting the config variable? Not sure if I'm reading this right, or what the intention is...

* Are `cleanup_interval_jobs` and `cleanup_interval_seconds` really both default to `nil`?  That's what the README says under https://github.com/bensheldon/good_job#configuration-options. But if so... does the automatic scheduled cleanupper actually run by default at all? It seems like if both are nil, it's not running, no?  Or are the docs wrong about both defaulting nil?

I'm still curious about whether this scheduled cleanup job is running multiple times in multiple processes if I'm running multiple worker procesess? And, especially if so, if it's possible to disable it?
